lib/semantics/mod-file.cc

```diff
@@ -52,43 +57,65 @@ static std::ostream &PutLower(std::ostream &, const Symbol &);
 static std::ostream &PutLower(std::ostream &, const DeclTypeSpec &);
 static std::ostream &PutLower(std::ostream &, const std::string &);
 static std::string CheckSum(const std::string &);
+static bool MakeReadonly(const std::string &);
 
 bool ModFileWriter::WriteAll() {
-  for (const auto &scope : Scope::globalScope.children()) {
-    if (scope.kind() == Scope::Kind::Module) {
-      auto &symbol{*scope.symbol()};  // symbol must be present for module
-      if (!symbol.test(Symbol::Flag::ModFile)) {
-        WriteOne(symbol);
-      }
-    }
-  }
+  WriteChildren(Scope::globalScope);
   return errors_.empty();
 }
 
-bool ModFileWriter::WriteOne(const Symbol &modSymbol) {
-  CHECK(modSymbol.has<ModuleDetails>());
-  auto name{parser::ToLowerCaseLetters(modSymbol.name().ToString())};
-  std::string path{ModFilePath(dir_, name)};
-  std::ofstream os{path};
-  PutSymbols(*modSymbol.scope());
-  std::string all{GetAsString(name)};
-  auto header{GetHeader(all)};
-  os << header << all;
-  os.close();
-  if (!os) {
-    errors_.emplace_back(
-        "Error writing %s: %s"_err_en_US, path.c_str(), std::strerror(errno));
-    return false;
+void ModFileWriter::WriteChildren(const Scope &scope) {
+  for (const auto &child : scope.children()) {
+    WriteOne(child);
+  }
+}
+
+void ModFileWriter::WriteOne(const Scope &scope) {
+  auto *symbol{scope.symbol()};
+  if (scope.kind() != Scope::Kind::Module) {
+    return;
   }
-  return true;
+  if (!symbol->test(Symbol::Flag::ModFile)) {
+    auto *ancestor{symbol->get<ModuleDetails>().ancestor()};
```

Comment:
The body of this `if` might make a good subroutine, even though it would be called only once, since the implementation of the writing of the module file is independent of the traversal context around it.

Reply:
Agreed